### ISSUE TITLE
[Schema] Versioning proposal

### ISSUE THREAD

**Issue by teoli2003** [mdn/browser-compat-data#107]
Hi all!

Experimentation showed that it is difficult to come with the perfect schema for browser compat from the start. As we expect the schema to evolve, we need a way to be sure that this evolution is possible.

The problems to solve are:
- As MDN (for each macro), and third-party tools, will use the data as soon as possible, having several (hopefully not more than 2!) coexisting versions of the schema is probable.
- The browser compat data will be cut in different files: we must be able to associate each of these files with a specific schema version.

Idea:
Start each file containing browser compat data with a version value
{"version": "1.0.0",
 "css" : {...}
}

Semantic versioning should be used (3rd digit: minor change, don't break compat, 2nd digit: upward compat, 1st digit: no compatibility). That way macros and 3rd-party can know the version used for a file.
As the schema is controlled here, we can ensure that the versioning is indeed semantic.

Note that I don't think allowing the coexistence of several version of the schema inside the same file is useful.

**Comment by jwhitlock** [comment]
I love the idea of versioning.  This can be combined with documentation that refers to the version number that a feature was added or removed, and schema validators can change behavior based on the given version number.

I am having a hard time thinking of a "3rd digit" change, that wouldn't be equivalent to the commit number changing.

**Comment by SebastianZ** [comment]
Generally, I think versioning is as good idea.

Two notes though:
1. Would it make sense to provide a (versioned) API rather than having to
maintain several versions of the data?
2. I also think that two digits for the version number should probably be
enough.

Sebastian

**Comment by Elchi3** [comment]
> 1. Would it make sense to provide a (versioned) API rather than having to maintain several versions of the data? 

At the moment we are far away from bringing back the compat data API. In the future, we hope to have an API, but for the moment I propose to concentrate on the data itself. This might mean intermediate solutions, but I think we shouldn't block ourselves until we are again in a position where we can work on an API.

> 2. I also think that two digits for the version number should probably be enough.

Likely.

Another idea to ensure our data is versioned, is to use git branches and/or git tags as opposed to putting version information into the data itself.

**Comment by teoli2003** [comment]
>I am having a hard time thinking of a "3rd digit" change, that wouldn't be equivalent to the commit number changing. Could you give some examples of changes that require version bumping?

You are right, the classical SemVer doesn't work well with Schema. But there is this: http://snowplowanalytics.com/blog/2014/05/13/introducing-schemaver-for-semantic-versioning-of-schemas/

It defines SemVer for Schemas:
Given a version number MODEL-REVISION-ADDITION, increment the:

    MODEL when you make a breaking schema change which will prevent interaction with any historical data
    REVISION when you make a schema change which may prevent interaction with some historical data
    ADDITION when you make a schema change that is compatible with all historical data

cc/ @jeremiepat as he have more experience than me on SemVer.

**Comment by JeremiePat** [comment]
I have no strong opinion and I see proper use case for the third digit. In my opinion we can use SemVer quite straight forward:

* 1st digit change: We introduced a breaking change 
  _meaning for a schema_: **Some old values no longer validate**
* 2nd digit change: We introduced some novelty that are backward compatible
  _meaning for a schema_: **Old values still validate but some new values are now allowed**
* 3rd digit change: Bug fix that allows to still validate old values without introducing new ones.
  _meaning for a schema_: **Some edge cased values that should have validate now properly validate** (note: _this could be considered 2nd digit change if the edge case isn't really at the edge_)**; we've optimized our schema to make it simpler and/or faster and/or more readable without introducing new values nor breaking existing ones, etc.**

By many aspects in the article suggested by @teoli2003 I don't really see the difference between MODEL and REVISION as in both cases it breaks compatibility with data. The trick with that is that many people do not expect a breaking change with a 2nd digit change which makes me skeptical about it.
